File: ailake-go/binary.go
// SPDX-License-Identifier: MIT OR Apache-2.0
// Copyright (c) 2026 Thiago Egon Lange
// Binary Hamming flat index: deserializer and brute-force search.
//
// Wire format: bincode v1 serialization of ailake_index::BinaryIndex:
//
//	codes:         Vec<u8>         (u64 length + flat packed-bit bytes)
//	bytes_per_vec: usize           (u64 LE — ceil(dim/8))
//	row_ids:       Vec<u64>        (u64 length + u64 values)
//	metric:        u32             (enum: 0=cosine, 1=euclidean, 2=dot, 3=normalized-cosine)
//	dim:           u32
//	raw_f16:       Option<Vec<u16>> (tag 0x00=None, 0x01=Some + u64 length + u16 values)
package ailake

import (
	"fmt"
	"math"
	"math/bits"
	"sort"
)

// BinaryIndex is a deserialized AI-Lake Binary Hamming flat index.
type BinaryIndex struct {
	// Packed bit codes, flat: entry i at Codes[i*BytesPerVec : (i+1)*BytesPerVec].
	Codes       []byte
	BytesPerVec int
	RowIDs      []uint64
	Metric      uint8
	Dim         uint32
	// Raw F16 vectors for reranking (nil when keep_raw=false).
	// Flat: entry i at RawF16[i*Dim : (i+1)*Dim].
	RawF16 []uint16
}

// DeserializeBinary deserializes a bincode-encoded BinaryIndex from buf.
func DeserializeBinary(buf []byte) (*BinaryIndex, error) {
	r := newBincodeReader(buf)

	// codes: Vec<u8>
	codesLen, err := r.readUsize()
	if err != nil {
		return nil, fmt.Errorf("binary: read codes length: %w", err)
	}
	codes, err := r.readN(int(codesLen))
	if err != nil {
		return nil, fmt.Errorf("binary: read codes: %w", err)
	}
	codesCopy := make([]byte, len(codes))
	copy(codesCopy, codes)

	// bytes_per_vec: usize
	bpv, err := r.readUsize()
	if err != nil {
		return nil, fmt.Errorf("binary: read bytes_per_vec: %w", err)
	}

	// row_ids: Vec<u64>
	rowIDs, err := r.readU64Slice()
	if err != nil {
		return nil, fmt.Errorf("binary: read row_ids: %w", err)
	}

	// metric: u32
	metricVariant, err := r.readU32()
	if err != nil {
		return nil, fmt.Errorf("binary: read metric: %w", err)
	}

	// dim: u32
	dim, err := r.readU32()
	if err != nil {
		return nil, fmt.Errorf("binary: read dim: %w", err)
	}

	// raw_f16: Option<Vec<u16>>
	optTag, err := r.readU8()
	if err != nil {
		return nil, fmt.Errorf("binary: read raw_f16 tag: %w", err)
	}
	var rawF16 []uint16
	if optTag == 1 {
		n, err := r.readUsize()
		if err != nil {
			return nil, fmt.Errorf("binary: read raw_f16 length: %w", err)
		}
		rawF16 = make([]uint16, n)
		for i := range rawF16 {
			v, err := r.readN(2)
			if err != nil {
				return nil, fmt.Errorf("binary: raw_f16[%d]: %w", i, err)
			}
			rawF16[i] = uint16(v[0]) | uint16(v[1])<<8
		}
	}

	return &BinaryIndex{
		Codes:       codesCopy,
		BytesPerVec: int(bpv),
		RowIDs:      rowIDs,
		Metric:      uint8(metricVariant),
		Dim:         dim,
		RawF16:      rawF16,
	}, nil
}

// hammingBinary counts differing bits between two equal-length byte slices.
// Uses uint64 chunks for throughput; bits.OnesCount64 maps to a single
// POPCNT instruction on x86_64 and VCNT+UADDLV on aarch64.
func hammingBinary(a, b []byte) int {
	n := len(a)
	if len(b) < n {
		n = len(b)
	}
	var total int
	chunks := n / 8
	for i := 0; i < chunks; i++ {
		base := i * 8
		var a64, b64 uint64
		for j := 0; j < 8; j++ {
			a64 |= uint64(a[base+j]) << (uint(j) * 8)
			b64 |= uint64(b[base+j]) << (uint(j) * 8)
		}
		total += bits.OnesCount64(a64 ^ b64)
	}
	for i := chunks * 8; i < n; i++ {
		total += bits.OnesCount8(a[i] ^ b[i])
	}
	return total
}

// f32ToBits binarizes a float32 vector: sign(x) ≥ 0 → bit 1, else bit 0.
// Bits packed MSB-first within each byte (bit 7 = dimension 0).
// Output length: ceil(len(v)/8) bytes.
func f32ToBits(v []float32) []byte {
	n := (len(v) + 7) / 8
	out := make([]byte, n)
	for i, val := range v {
		if val >= 0.0 {
			out[i/8] |= 0x80 >> uint(i%8)
		}
	}
	return out
}

// Search returns the top-k (rowID, distance) pairs sorted ascending by Hamming distance.
// rerank: if true and RawF16 is available, the top 3×k Hamming-nearest candidates are
// reranked using exact F16 distances with the configured metric.
func (idx *BinaryIndex) Search(query []float32, topK int, rerank bool) []SearchResult {
	if len(idx.RowIDs) == 0 || topK <= 0 {
		return nil
	}
	n := len(idx.RowIDs)
	bpv := idx.BytesPerVec
	dim := int(idx.Dim)

	qBits := f32ToBits(query)

	type scored struct {
		i    int
		dist float32
	}

	all := make([]scored, n)
	for i := 0; i < n; i++ {
		code := idx.Codes[i*bpv : (i+1)*bpv]
		h := hammingBinary(qBits, code)
		all[i] = scored{i, float32(h)}
	}

	sort.Slice(all, func(a, b int) bool { return all[a].dist < all[b].dist })

	candidates := topK
	if rerank && len(idx.RawF16) > 0 {
		candidates = topK * 3
	}
	if candidates > n {
		candidates = n
	}
	top := all[:candidates]

	if rerank && len(idx.RawF16) > 0 {
		reranked := make([]scored, len(top))
		for j, s := range top {
			dbF16 := idx.RawF16[s.i*dim : (s.i+1)*dim]
			dbF32 := make([]float32, dim)
			for k, bits := range dbF16 {
				dbF32[k] = f16ToF32Go(bits)
			}
			var dist float32
			switch idx.Metric {
			case MetricCosine, MetricNormalizedCosine:
				dist = cosineDistGo(query, dbF32)
			case MetricDotProduct:
				dist = dotDistGo(query, dbF32)
			default:
				dist = euclideanDistGo(query, dbF32)
			}
			reranked[j] = scored{s.i, dist}
		}
		sort.Slice(reranked, func(a, b int) bool { return reranked[a].dist < reranked[b].dist })
		top = reranked
	}

	if topK > len(top) {
		topK = len(top)
	}
	out := make([]SearchResult, topK)
	for i, s := range top[:topK] {
		out[i] = SearchResult{RowID: idx.RowIDs[s.i], Distance: s.dist}
	}
	return out
}

// hammingDistance is the public Hamming distance function (used in tests).
func hammingDistance(a, b []byte) int { return hammingBinary(a, b) }

// ensure math import used
var _ = math.Sqrt

File: ailake-index/src/binary.rs
// SPDX-License-Identifier: MIT OR Apache-2.0
//! Hamming flat index: brute-force search over binary-quantized vectors.
//!
//! Use case: embedding models that produce binary-compatible vectors
//! (e.g. Cohere embed-v3 binary mode, Jina ColBERT). Sign of each float
//! dimension maps to one bit; distance = Hamming(a, b) = popcount(a XOR b).
//!
//! 32× compression vs F32 (1 bit/dim vs 32 bits/dim).
//! For float embeddings with random rotation use RaBitQ instead — it achieves
//! much better recall by applying an orthogonal projection before binarization.
//!
//! Search:
//!   1. Binarize query (sign → bit, once per search).
//!   2. For each stored vector: Hamming(q_bits, code) via AVX2/NEON/scalar.
//!   3. Partial select top `candidates = rerank_factor × k`, sort.
//!   4. Optional reranking: exact F16 distance for top candidates.

use ailake_core::{AilakeError, AilakeResult, RowId, VectorMetric};
use ailake_vec::{binary_quant, exact_distance};
use half::f16;
use serde::{Deserialize, Serialize};

// ── Config ────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BinaryConfig {
    /// Keep raw F16 vectors for exact reranking. Default true.
    #[serde(default = "default_keep_raw")]
    pub keep_raw: bool,
}

fn default_keep_raw() -> bool {
    true
}

impl Default for BinaryConfig {
    fn default() -> Self {
        Self { keep_raw: true }
    }
}

// ── Index ─────────────────────────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize)]
pub struct BinaryIndex {
    /// Bit-packed codes, flat: entry `i` occupies bytes `[i*bpv..(i+1)*bpv]`.
    pub codes: Vec<u8>,
    /// Bytes per vector = `ceil(dim / 8)`.
    pub bytes_per_vec: usize,
    pub row_ids: Vec<u64>,
    pub metric: VectorMetric,
    pub dim: u32,
    /// Raw F16 vectors for reranking, stored flat (dim values per entry).
    pub raw_f16: Option<Vec<f16>>,
}

impl BinaryIndex {
    /// Build from raw F32 vectors.
    pub fn build(
        row_ids: &[RowId],
        vectors: &[Vec<f32>],
        metric: VectorMetric,
        keep_raw: bool,
    ) -> AilakeResult<Self> {
        if vectors.is_empty() {
            return Err(AilakeError::InvalidArgument(
                "BinaryIndex::build requires at least one vector".into(),
            ));
        }
        let dim = vectors[0].len();
        let bytes_per_vec = (dim + 7) / 8;

        // Encode all vectors to packed bits
        let mut codes = Vec::with_capacity(vectors.len() * bytes_per_vec);
        for v in vectors {
            let bits = binary_quant::f32_to_bits(v);
            codes.extend_from_slice(&bits);
        }

        let raw_f16 = if keep_raw {
            Some(
                vectors
                    .iter()
                    .flat_map(|v| v.iter().map(|&x| f16::from_f32(x)))
                    .collect(),
            )
        } else {
            None
        };

        Ok(Self {
            codes,
            bytes_per_vec,
            row_ids: row_ids.iter().map(|r| r.as_u64()).collect(),
            metric,
            dim: dim as u32,
            raw_f16,
        })
    }

    pub fn node_count(&self) -> u64 {
        self.row_ids.len() as u64
    }

    /// Search for the `top_k` nearest vectors using Hamming distance.
    ///
    /// If `rerank_factor` > 1 and raw F16 vectors are stored, the top
    /// `rerank_factor × top_k` Hamming-nearest candidates are reranked
    /// with exact F16 distances using the table's configured metric.
    ///
    /// Returns `(RowId, distance)` pairs sorted ascending by distance.
    pub fn search(
        &self,
        query: &[f32],
        top_k: usize,
        rerank_factor: Option<usize>,
    ) -> Vec<(RowId, f32)> {
        if self.row_ids.is_empty() {
            return vec![];
        }
        let n = self.row_ids.len();
        let bpv = self.bytes_per_vec;

        let q_bits = binary_quant::f32_to_bits(query);

        // Coarse scan: Hamming distance for every entry (sequential — shard
        // parallelism handled externally by SearchSession)
        let mut scored: Vec<(usize, f32)> = self
            .codes
            .chunks_exact(bpv)
            .enumerate()
            .map(|(i, code)| {
                let d = binary_quant::hamming_distance(&q_bits, code) as f32;
                (i, d)
            })
            .collect();

        let candidates = rerank_factor.unwrap_or(1).max(1) * top_k;
        let cmp = |a: &(usize, f32), b: &(usize, f32)| {
            a.1.partial_cmp(&b.1).unwrap_or(std::cmp::Ordering::Equal)
        };
        if candidates < n {
            scored.select_nth_unstable_by(candidates - 1, cmp);
            scored.truncate(candidates);
        }
        scored.sort_unstable_by(cmp);

        if let Some(ref raw) = self.raw_f16 {
            let dim = self.dim as usize;
            let mut reranked: Vec<(usize, f32)> = scored
                .iter()
                .map(|&(i, _)| {
                    let db_f16 = &raw[i * dim..(i + 1) * dim];
                    let db_f32: Vec<f32> = db_f16.iter().map(|x| x.to_f32()).collect();
                    let d = exact_distance(self.metric, query, &db_f32);
                    (i, d)
                })
                .collect();
            reranked.sort_unstable_by(cmp);
            reranked
                .into_iter()
                .take(top_k)
                .map(|(i, d)| (RowId::new(self.row_ids[i]), d))
                .collect()
        } else {
            scored
                .into_iter()
                .take(top_k)
                .map(|(i, d)| (RowId::new(self.row_ids[i]), d))
                .collect()
        }
    }
}

// ── Serializer ────────────────────────────────────────────────────────────────

pub struct BinarySerializer;

impl BinarySerializer {
    pub fn to_bytes(index: &BinaryIndex) -> AilakeResult<Vec<u8>> {
        bincode::serialize(index).map_err(|e| AilakeError::Bincode(e.to_string()))
    }

    pub fn from_bytes(bytes: &[u8]) -> AilakeResult<BinaryIndex> {
        bincode::deserialize(bytes).map_err(|e| AilakeError::Bincode(e.to_string()))
    }
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use rand::{rngs::StdRng, Rng, SeedableRng};

    fn binary_vecs(n: usize, dim: usize, seed: u64) -> Vec<Vec<f32>> {
        let mut rng = StdRng::seed_from_u64(seed);
        (0..n)
            .map(|_| (0..dim).map(|_| if rng.gen::<bool>() { 1.0f32 } else { -1.0 }).collect())
            .collect()
    }

    #[test]
    fn exact_self_is_top1() {
        let vecs = binary_vecs(50, 64, 1);
        let row_ids: Vec<RowId> = (0..50u64).map(RowId::new).collect();
        let idx =
            BinaryIndex::build(&row_ids, &vecs, VectorMetric::Cosine, false).unwrap();

        let query = vecs[7].clone();
        let results = idx.search(&query, 1, None);
        assert_eq!(results.len(), 1);
        assert_eq!(results[0].0, RowId::new(7));
    }

    #[test]
    fn rerank_produces_exact_distances() {
        let vecs = binary_vecs(100, 64, 99);
        let row_ids: Vec<RowId> = (0..100u64).map(RowId::new).collect();
        let idx =
            BinaryIndex::build(&row_ids, &vecs, VectorMetric::Cosine, true).unwrap();

        let query = vecs[0].clone();
        let results = idx.search(&query, 5, Some(3));
        // All distances should be valid floats in [0.0, 2.0] for cosine
        for (_, d) in &results {
            assert!(d.is_finite() && *d >= 0.0, "invalid distance {d}");
        }
    }

    #[test]
    fn serialization_roundtrip() {
        let vecs = binary_vecs(30, 32, 42);
        let row_ids: Vec<RowId> = (0..30u64).map(RowId::new).collect();
        let idx =
            BinaryIndex::build(&row_ids, &vecs, VectorMetric::Euclidean, false).unwrap();

        let bytes = BinarySerializer::to_bytes(&idx).unwrap();
        let idx2 = BinarySerializer::from_bytes(&bytes).unwrap();
        let q = vecs[0].clone();
        let r1 = idx.search(&q, 1, None);
        let r2 = idx2.search(&q, 1, None);
        assert_eq!(r1[0].0, r2[0].0);
    }

    #[test]
    fn node_count_matches() {
        let vecs = binary_vecs(20, 16, 7);
        let row_ids: Vec<RowId> = (0..20u64).map(RowId::new).collect();
        let idx = BinaryIndex::build(&row_ids, &vecs, VectorMetric::Cosine, false).unwrap();
        assert_eq!(idx.node_count(), 20);
    }
}
